README.md
# homeR

`homeR` is a collection of functions that we have
found useful at Neurobat in the analysis of experiments on buildings.

## Installation

Get the released version from CRAN:

```r
install.packages("homeR")
```

Or the development version from GitHub:

```r
# install.packages("devtools")
devtools::install_github("neurobat/homeR")
```

## Building the package

We recommend using RStudio to build this package. Make sure you have the latest
version installed (I'm currently using 0.99.896), and make also sure you have
`roxygen2` and `testthat` installed:

```r
install.packages("roxygen2")
install.packages("testthat")
```

From RStudio, create a new project (File -> New Project -> Existing Directory) and give
the project directory. Configure RStudio to build the documentation from Roxygen:
Build -> Configure Build Tools -> Generate documentation with Roxygen. Under
"Configure" ensure that "Rd files" and "NAMESPACE file" are checked.

Verify that you can build the package by running Document (Shift-Ctrl/Cmd-D) and
Test Package (Shift-Ctrl/Cmd-T).

## Committing generated files

The `NAMESPACE` and `*.Rd` files, although generated from the sources, should be
committed and kept under Git so that anyone cheking out the project from GitHub
can build and install it from the command line.

NEWS.md
# homeR 0.3.0

## New features

* New `vcov.bhm()` function returns the covariance matrix estimated at the posterior's mode.
* New `logposterior()` function returns the log-posterior function for the model's data.
* User can specify a fixed base load.


# homeR 0.2

## New features

* New `bhm()` function estimates a building's heating coefficients (total heat
loss coefficient, base temperature, etc).


tests/testthat.R
library(testthat)
library(homeR)

test_check("homeR") 


tests/testthat/test_bhm.R
library(homeR)

library(plyr)
context("Bayesian heating model")

# The heating energy model assumes that the heating energy grows linearly as
# the outdoor temperature decreases, but remains constant above the base temperature.
# For unit tests we use a model with base temp = 1, K = 1, DHW = 1 and sigma = 1e-2.
# 
#        ^
#        |
#        |               |
#        |      \
#        |       \
#        |        \      |
# Energy |         \
#        |          \
#        |           \   |
#        |            \
#        | total heat loss
#        |    coef (K)  \|
#        |             ( \---------------
#        +--------------------------------->
#                        |          Outdoor
#                      base          temp.
#                      temp.

set.seed(1111)

K <- 1
tb <- 1
DHW <- 1
sigma <- 1e-2

temps <- tb + c(-2, -1, 0, 1)
E <- K * pmax(tb - temps, 0) + DHW + rnorm(length(temps), 0, sigma)

fourDayData <- data.frame(E = E, T = temps)

expect_equal_coefs <- function(model)
  expect_equal(coef(model), c(K = K, tb = tb, DHW = DHW, sigma = sigma), tolerance = 1e-2, scale = 1)

expect_within <- function(object, expected, pm)
  expect_equal(as.vector(object), expected, tolerance = pm, scale = 1)

test_that("finds coefs on four-day data", {
  model <- bhm(E ~ T, fourDayData)
  expect_equal_coefs(model)
  expect_lt(sum(resid(model)^2), 1e-1)
})

test_that("negative energy yields an error", {
  dataWithNegativeEnergy <- transform(fourDayData, E = -E)
  expect_error(bhm(E ~ T, dataWithNegativeEnergy))
})

test_that("provides residuals", {
  model <- bhm(E ~ T, fourDayData)
  expect_length(residuals(model), nrow(fourDayData))
})

test_that("find coefs on four times two-day data", {
  fourTwoDayEpochData <- adply(fourDayData, 1,
                               summarise, E = 2 * E, T = I(list(c(T, T))))
  model <- bhm(E ~ T, fourTwoDayEpochData)
  expect_equal_coefs(model)
})

test_that("works with synthetic data from paper", {
  load("fakeMonthlyEnergy.RData")
  model <- bhm(Energy ~ DailyMeans, fakeMonthlyEnergy)
  coefs <- coef(model)
  stderr <- sqrt(diag(vcov(model)))
  expect_within(coefs['K'], 20, pm = 5e-2)
  expect_within(stderr['K'], 0.1, pm = 0.1)
  expect_within(coefs['tb'], 12, pm = 5e-2)
  expect_within(stderr['tb'], 0.2, pm = 0.02)
  expect_within(coefs['DHW'], 100, pm = 2)
  expect_within(stderr['DHW'], 2, pm = 1)
  expect_within(coefs['sigma'], 30, pm = 1)
  expect_within(stderr['sigma'], 3, pm = 1)
})

test_that("works with heat counter data", {
  load("alv.RData") # heat counter data from paper
  model <- bhm(Energy ~ DailyMeans, alv)
  coefs <- coef(model)
  expect_within(coefs['K'], 2.4, pm = 0.1)
  expect_within(coefs['tb'], 17.7, pm = 0.1)
  expect_within(coefs['DHW'], 0.7, pm = 0.1)
  expect_within(coefs['sigma'], 7.8, pm = 0.1)
})

test_that("provided log-posterior has maximum at estimated coefficients", {
  model <- bhm(E ~ T, fourDayData)
  coefs <- coef(model)
  nonOptimalCoefs <- lapply(coefs, function(x) jitter(rep(x, 2)))
  logp <- logposterior(model)
  logPosteriorMode <- do.call(logp, as.list(coefs))
  logPosteriorElsewhere <- do.call(logp, nonOptimalCoefs)
  expect_true(all(logPosteriorElsewhere < logPosteriorMode))
  # with a small jitter, the logp should never decrease by more than, say, 20
  expect_true(all(logPosteriorElsewhere > logPosteriorMode - 20))
})

test_that("user can specify an arbitrary DHW", {
  threeDaysBelowBaseTemp <- fourDayData[1:3, ]
  model <- bhm(E ~ T, data = threeDaysBelowBaseTemp, baseLoad = 0)
  coefs <- coef(model)
  expect_within(coefs['K'], K, pm = 0.1)
  expect_within(coefs['DHW'], 0, pm = 0.1)
})

tests/testthat/test_pmv.R
library(homeR)
context("Predicted Mean Vote")

test_that("pmv() handles scalars", {
  vote <- pmv(clo = 1, met = 1.2, air.temp = 19, saturation = 40)
  expect_gt(vote, -0.6)
  expect_lt(vote, -0.5)
})

test_that("pmv() handles vectors", {
  votes <- pmv(clo = 1, met = 1.2, air.temp = c(19, 30), sat = 40)
  expect_length(votes, 2)
  expect_equal(votes, c(pmv(clo = 1, met = 1.2, air.temp = 19, sat = 40), pmv(clo = 1, 
    met = 1.2, air.temp = 30, sat = 40)))
}) 


R/bhm.R
#' Bayesian Heating Model
#' 
#' Estimates the parameters of a building's heating model.
#' 
#' \code{bhm} assumes that the heating energy for a building has been measured
#' over several time periods (not necessarily of equal length). The \code{data}
#' data frame should have one row per measurement period. The energy vector (whose
#' name is given on the left-hand side of the formula) will have the total energy
#' measured during each period. The daily temperature vector (whose name is given on 
#' the right-hand side of the formula) will have either a vector of average daily
#' temperatures (when each measurement period is just one day) or a list of vectors
#' (when each measurement period can be an arbitrary number of days).
#' 
#' @param formula an object of class "\link{formula}": a description of which variable
#' holds the energy readouts and which variable holds the daily temperatures.
#' @param data a data frame in which the energy and daily temperatures are to be found.
#' @param baseLoad a optional constant base load, e.g. for domestic hot water preparation.
#' @return \code{bhm} returns an object of \link{class} "\code{bhm}". The generic
#' accessor functions \code{coefficients}, \code{vcov} and \code{residuals} extract the usual
#' information from the fitted model, while \code{logposterior} will return a function
#' that evaluates the log-posterior as a function of the parameters.
#' @examples 
#' set.seed(1111)
#' 
#' # Simple, but unrealistic parameters
#' K <- 1
#' tb <- 1
#' DHW <- 1
#' sigma <- 1e-2
#' temps <- tb + c(-2, -1, 0, 1)
#' 
#' # With daily measurements
#' E <- K * pmax(tb - temps, 0) + DHW + rnorm(length(temps), 0, sigma)
#' fourDayData <- data.frame(E = E, T = temps)
#' fourDayData
#' \dontrun{
#' fit <- bhm(E ~ T, fourDayData)
#' coef(fit)
#' resid(fit)
#' }
#' 
#' # With two-day measurements
#' fourTimesTwoDayData <- with(fourDayData,
#'                             data.frame(E = 2 * E,
#'                             T = I(lapply(T, function(x) c(x, x)))))
#' fit2 <- bhm(E ~ T, fourTimesTwoDayData)
#' coef(fit2)
#' resid(fit2)
#' @export
bhm <- function(formula, data, baseLoad = NULL) {
  Q <- energy(formula, data)
  stopifnot(all(Q >= 0))
  T <- temperatures(formula, data)
  result <- list()
  result$logPosterior <- logPosteriorFunction(Q, T)
  fit <- posteriorMode(result$logPosterior, baseLoad)
  stopifnot('DHW' %in% names(fit$par) || !is.null(baseLoad))
  result$coefficients <- c(fit$par, DHW = baseLoad)
  result$residuals <- with(as.list(result$coefficients),
                           Q - epochEnergy(K, tb, DHW, T))
  result$vcov <- solve(fit$hessian)
  class(result) <- "bhm"
  result
}

energy <- function(formula, data) {
  terms <- terms(formula)
  eval(attr(terms, "variables"), envir = data)[[attr(terms, "response")]]
}

temperatures <- function(formula, data) {
  terms <- terms(formula)
  data[[attr(terms, "term.labels")]]
}

posteriorMode <- function(logposterior, baseLoad = NULL) {
  obj <- function(par, ...) {
   - do.call(logposterior, as.list(c(par, ...)))
  }
  start <- list(K = 1, tb = 10, DHW = 1, sigma = 1)
  fixed <- if (!is.null(baseLoad)) list(DHW = baseLoad) else list()
  start[names(fixed)] <- NULL
  fit <- stats::optim(start, obj, gr = NULL, fixed, method = "SANN", control = list(maxit = 20000))
  fit <- stats::optim(fit$par, obj, gr = NULL, fixed, control = list(maxit = 1000), hessian = TRUE)
  fit
}

logPosteriorFunction <- function(energy, temperatures) {
  mlp <- function(K = 10, tb = 20, DHW = 0, sigma = 100) {
    if (sigma <= 0) return(-Inf)
    predictedEnergy <- epochEnergy(K, tb, DHW, temperatures)
    stopifnot(length(predictedEnergy) == length(energy))
    epochLength <- sapply(temperatures, length)
    - ((1 + length(energy)) * log(sigma) + 
         chisquare(energy, predictedEnergy, sqrt(epochLength) * sigma) / 2)
  }
  Vectorize(mlp)
}

chisquare <- function(energy, predictedEnergy, epochSigma) {
  stopifnot(length(energy) == length(predictedEnergy))
  stopifnot(length(energy) == length(epochSigma))
  sum(((energy - predictedEnergy) / epochSigma) ^ 2)
}

epochEnergy <- function(K, tb, DHW, temperatures) {
  result <- sapply(temperatures, function(temperature) sum(dailyEnergy(K, tb, DHW, temperature)))
  stopifnot(length(result) == length(temperatures))
  result
}

dailyEnergy <- function(K, tb, DHW, temperature) K * pos(tb - temperature) + DHW

pos <- function(x) pmax(x, 0)

#' @export
residuals.bhm <- function(object, ...) object$residuals

#' @export
vcov.bhm <- function(object, ...) object$vcov

#' Log-posterior of a Bayesian Heating Model
#' 
#' Provides the log-posterior of a heating model given the data, as a function
#' of the model's parameters.
#' 
#' @param bhm a fitted model returned by a call to \code{bhm()}
#' @return a function of the model's parameters (currently K, tb, DHW and sigma)
#' @export
logposterior <- function(bhm) bhm$logPosterior

R/utils.R
#' Positive Part
#' 
#' Computes the positive part of each element of a vector.
#' 
#' @param x A numeric vector
#' @return The argument with all negative entries replaced by 0
pos <- function(x) pmax(x, 0)

#' Daily Energy Model
#' 
#' Computes the daily energy model without the noise component.
#' 
#' @param dailyMeans A vector of mean daily temperatures
#' @param K The building's total heat loss coefficient
#' @param baseTemp The building's base temperature
#' @param DHW The building's average base load
#' @return A vector of predicted energy consumption for each day.
dailyEnergyModel <- function(dailyMeans, K = 20, baseTemp = 12, DHW = 100)
  K * pos(baseTemp - dailyMeans) + DHW

#' Period Energy Model
#' 
#' Computes the energy consumption over several days, without the noise component.
#' 
#' @inheritParams dailyEnergyModel
#' @return The total predicted energy consumption for that period.
periodEnergyModel <- function(...) sum(dailyEnergyModel(...))

#' Chi-square for a single datum
#' 
#' Computes the chi-square statistic for a single reporting period.
#' 
#' @param response The value predicted by the model given the data and the parameters
#' @param model The model that predicts the response given data and parameters
#' @param sigma The standard deviation that should be used for computing the Chi-square
#' @param ... Extra arguments passed to the model
#' @return The chi-square for that single datum.
chisquare1 <- function(response, model, sigma, ...) {
  ((model(...) - response)/sigma)^2
}

#' Chi-square for several reporting periods
#' 
#' @param periods A data frame with two elements: `Energy`, a vector of the energy
#' recorded during each period, and `DailyMeans`, a list of mean daily temperature
#' vectors for each reporting period.
#' @param U, baseTemp, DHW, sigma As above
chisquare <- function(periods, U, baseTemp, DHW, sigma) {
  stopifnot("Energy" %in% names(periods))
  stopifnot("DailyMeans" %in% names(periods))
  chisquares <- apply(periods, MARGIN = 1, function(period) chisquare1(period$Energy, 
    periodEnergyModel, sqrt(length(period$DailyMeans)) * sigma, 
    period$DailyMeans, U, baseTemp, DHW))
  sum(chisquares)
}

#' The log-likelihood of the data given the parameters
#' 
#' @param periods, U, baseTemp, DHW, sigma As above
loglikelihood <- function(periods, U, baseTemp, DHW, sigma) {
  if (sigma <= 0) 
    return(-Inf)
  if (DHW < 0) 
    return(-Inf)
  -(1 + nrow(periods)) * log(sigma) - chisquare(periods, U, baseTemp, 
    DHW, sigma)/2
}
loglikelihood <- Vectorize(loglikelihood, c("U", "baseTemp", "DHW", 
  "sigma"))

#' Estimate building parameters and their uncertainties from the data
#' 
#' @param periods The energy and mean outdoor temperature, as above
#' @return A list of two elements: `params`, a list of estimated parameters, 
#' and `covariance`, the covariance matrix
estimateParameters <- function(periods) {
  # Find parameters that maximise log-likelihood
  objective <- function(params, data) {
    curried <- function(...) -loglikelihood(data, ...)
    do.call(curried, as.list(params))
  }
  optimal <- optim(par = c(10, 20, 0, 100), fn = objective, gr = NULL, 
    periods, hessian = TRUE)  # for covariance matrix
  params = list(U = 0, baseTemp = 0, DHW = 0, sigma = 0)
  params[] <- optimal$par
  list(params = params, covariance = solve(optimal$hessian))
} 


R/pmv.R
#' Predicted Mean Vote
#' 
#' Computes Fanger's predicted mean vote
#' 
#' Compute the predicted mean vote for one or more combinations of 
#' clo, met, air temperature and moisture saturation. The inputs arguments 
#' can be scalars or vectors.
#' 
#' @param clo Thermal insulation of clothing in [clo] (underwear, blouse/shirt, slacks/trousers, jacket, socks and shoes are approximately 1 clo)
#' @param met Physical activity in [met] (one person seated at rest is approximately 1 met)
#' @param air.temp Indoor air temperature (assumed equal to mean radiant temperature) in [C]
#' @param saturation Ratio of moisture content to moisture content of saturated air at the same temperature, in [\%] (approximately the same thing as relative humidity)
#' @return The predicted mean vote, a value between -3 (cold) to +3 (hot)
#' @references CIBSE Guide A, section 1.4 and 1.A1.2 (from which this implementation is derived)
#' @examples 
#' # With scalars
#' pmv(clo=1.0,
#'     met=1.2,
#'     air.temp=19,
#'     saturation=40)
#' # With vectors
#' pmv(clo=c(1.0, 1.5),
#'     met=c(1.2, 0.6),
#'     air.temp=c(19, 30),
#'     sat=c(35, 40))
#' @export
pmv <- function(clo, met, air.temp, saturation) {
  # Adapted from CIBSE Guide A, 1.A1.2, p. 51
  saturated.vapour.pressure <- exp(16.6536-4030.183/(air.temp+235)) # kPa
  vapour.pressure <- saturated.vapour.pressure*10*saturation # Pa
  clothing.insulation <- 0.155*clo
  met.rate <- met*58.15
  external.work <- 0 # simplification
  int.heat.prod <- met.rate - external.work
  clo.area.fact <- ifelse(clothing.insulation<0.078,
                          1 + 1.29*clothing.insulation,
                          1.05 + 0.645*clothing.insulation)
  convective.heat <- 0 # assumes null air velocity
  air.temp.kel <- air.temp + 273
  mean.rad.temp <- air.temp.kel # simplification
  # compute clothing surface temperature by iteration
  clo.surf.temp <- air.temp.kel +
    (35.5 - air.temp)/(3.5*(6.45*clothing.insulation + 0.1))
  p1 <- clothing.insulation*clo.area.fact
  p2 <- p1*3.96
  p3 <- p1*100 # Error in CIBSE Guide
  p4 <- p1*air.temp.kel
  p5 <- 308.7 - 0.028*int.heat.prod + p2*(mean.rad.temp/100)^4
  xn <- clo.surf.temp/100
  xf <- xn
  eps <- 0.000015
  hc <- 0
  repeat {
    xf <- (xf + xn)/2
    heat.transf.nat <- 2.38*abs(100*xf-air.temp.kel)^0.25
    hc <- pmax(convective.heat, heat.transf.nat)
    xn <- (p5 + p4*hc - p2*xf^4)/(100 + p3*hc)
    if (all(abs(xn-xf)<eps)) break
  }
  clo.surf.temp <- 100*xn - 273
  # end iteration
  heat.loss.skin <- 3.05*0.001*(5733-6.99*int.heat.prod-vapour.pressure) # CIBSE Guide has an error here
  heat.loss.sweat <- pmax(0.42*(int.heat.prod - 58.15), 0)
  heat.loss.respiration <- 1.7*0.00001*met.rate*(5867 - vapour.pressure)
  heat.loss.dry <- 0.0014*met.rate*(34 - air.temp)
  heat.loss.rad <- 3.96*clo.area.fact*(xn^4 - (mean.rad.temp/100)^4)
  heat.loss.conv <- clo.area.fact*hc*(clo.surf.temp - air.temp)
  therm.sens <- 0.303*exp(-0.036*met.rate) + 0.028
  therm.sens*(int.heat.prod - heat.loss.skin - heat.loss.sweat
              - heat.loss.respiration - heat.loss.dry
              - heat.loss.rad - heat.loss.conv)
}
pmv <- Vectorize(pmv)
